Resources/js/router.ts

```diff
@@ -0,0 +1,348 @@
+export interface RouteDefaults {
+  [index: string]: string;
+}
+
+export interface RouteRequirements {
+  [index: string]: string;
+}
+
+export interface RouteParams {
+  [index: string]: any;
+}
+
+export interface QueryParamAddFunction {
+  (prefix: string, params: any): void;
+}
+
+export interface Route {
+  tokens: ([string, string] | [string, string, string] | [string, string, string, string] | [string, string, string, string, boolean])[];
```

Comment:
For typing arrays of `['token1']`, `['token1', 'token2']`, `['token1', 'token2', 'token3']`, `['token1', 'token2', 'token3', 'token4']` and `['token1', 'token2', 'token3', 'token4', boolean]`. 

See [Tuples](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types)

Reply:
Reverted to `(string|boolean)[][]` because TypeScript automatically sees `(string|boolean)[][]` for `tokens` when importing the JSON dump file. 

Added a test for that and adapted the source code.